rust/routee-compass-powertrain/src/routee/prediction/interpolation/utils.rs

```diff
@@ -0,0 +1,109 @@
+use ordered_float::OrderedFloat;
+
+pub fn linspace(x0: f64, xend: f64, n: usize) -> Vec<f64> {
+    let dx = (xend - x0) / ((n - 1) as f64);
+    let mut x = vec![x0; n];
+    for i in 1..n {
+        x[i] = x[i - 1] + dx;
+    }
+    x
+}
+
+fn find_nearest_index(arr: &[OrderedFloat<f64>], target: OrderedFloat<f64>) -> usize {
+    let mut low = 0;
+    let mut high = arr.len() - 1;
+
+    while low < high {
+        let mid = low + (high - low) / 2;
+
+        if arr[mid] >= target {
+            high = mid;
+        } else {
+            low = mid + 1;
+        }
+    }
+
+    if low > 0 && arr[low] >= target {
+        low - 1
+    } else {
+        low
+    }
+}
+
+pub struct BilinearInterp {
+    pub x: Vec<OrderedFloat<f64>>,
+    pub y: Vec<OrderedFloat<f64>>,
+    pub values: Vec<Vec<f64>>,
+}
+
+impl BilinearInterp {
+    pub fn new(x: Vec<f64>, y: Vec<f64>, values: Vec<Vec<f64>>) -> Result<Self, String> {
+        if x.len() != values.len() {
+            return Err("Supplied `x` must have same dimensionality as `values`".to_string());
+        }
+        if y.len() != values[0].len() {
+            return Err("Supplied `y` must have same dimensionality as `values`".to_string());
+        }
+        if !x.windows(2).all(|w| w[0] < w[1]) {
+            return Err("Supplied `x` coordinates must be sorted and non-repeating".to_string());
+        }
+        if !y.windows(2).all(|w| w[0] < w[1]) {
+            return Err("Supplied `y` coordinates must be sorted and non-repeating".to_string());
+        }
+        let x = x.into_iter().map(OrderedFloat::from).collect();
+        let y = y.into_iter().map(OrderedFloat::from).collect();
+        Ok(BilinearInterp { x, y, values })
+    }
+
+    pub fn interpolate(&self, x: f64, y: f64) -> Result<f64, &'static str> {
+        let x_index = find_nearest_index(&self.x, OrderedFloat(x));
+        let y_index = find_nearest_index(&self.y, OrderedFloat(y));
+
+        if x_index >= self.x.len() - 1 || y_index >= self.y.len() - 1 {
+            return Err("Cannot interpolate outside of grid bounds");
+        }
+        let x0 = self.x[x_index].into_inner();
+        let x1 = self.x[x_index + 1].into_inner();
+        let y0 = self.y[y_index].into_inner();
+        let y1 = self.y[y_index + 1].into_inner();
+
+        let q11 = self.values[x_index][y_index];
+        let q12 = self.values[x_index][y_index + 1];
+        let q21 = self.values[x_index + 1][y_index];
+        let q22 = self.values[x_index + 1][y_index + 1];
+
+        let fxy1 = (x1 - x) / (x1 - x0) * q11 + (x - x0) / (x1 - x0) * q21;
```

Comment:
again, wondering if this is code comes from somewhere else, though i get it if it's just common sense that this is what linear interpolation is.

Reply:
Yeah this is based on [one formula for bilinear interpolation](https://en.wikipedia.org/wiki/Bilinear_interpolation). I'll add in the link so we remember where we got it from